File: browser/src/ordering.ts
export type ApplicationStep =
  | "navigate"
  | "preflight_morph"
  | "morph"
  | "validate_no_render"
  | "commit_snapshot_and_revision"
  | "reconcile_models_and_validation"
  | "restore_focus"
  | "queue_child_deliveries"
  | "reflect_url"
  | "dispatch_events"
  | "run_registered_effects"
  | "settle_feedback"
  | "retain_dom"
  | "request_fresh_render_without_replay"
  | "request_fresh_island"
  | "stop_live";

export interface ApplicationPlanInput {
  readonly protocol: 1 | 2;
  readonly outcome: "accepted" | "duplicate" | "rejected" | "refresh_required" | "fatal";
  readonly render: "redirect" | "navigated" | "html" | "no_render" | "none";
  readonly morph: "not_attempted" | "succeeded" | "failed_after_acceptance";
  readonly hasChildDeliveries: boolean;
  readonly hasReflectedUrl: boolean;
  readonly recovery:
    "retain_dom" | "retry" | "refresh_island" | "remount_island" | "navigate" | "stop" | null;
}

export function applicationPlan(
  render: "redirect" | "html" | "no_render",
  morph: "not_attempted" | "succeeded" | "failed_after_acceptance",
): readonly ApplicationStep[] {
  return applicationPlanV2({
    protocol: 1,
    outcome: "accepted",
    render,
    morph,
    hasChildDeliveries: false,
    hasReflectedUrl: false,
    recovery: render === "redirect" ? "navigate" : null,
  });
}

export function applicationPlanV2(input: ApplicationPlanInput): readonly ApplicationStep[] {
  if (input.render === "redirect" || input.render === "navigated") return ["navigate"];
  if (input.outcome === "rejected") return ["retain_dom", "settle_feedback"];
  if (input.outcome === "refresh_required") {
    return input.recovery === "navigate"
      ? ["navigate"]
      : ["retain_dom", "request_fresh_island", "settle_feedback"];
  }
  if (input.outcome === "fatal") {
    return input.recovery === "navigate"
      ? ["navigate"]
      : ["retain_dom", "stop_live", "settle_feedback"];
  }
  if (input.render === "html" && input.morph === "failed_after_acceptance") {
    return ["preflight_morph", "morph", "request_fresh_render_without_replay"];
  }

  const tail: ApplicationStep[] = [
    "commit_snapshot_and_revision",
    "reconcile_models_and_validation",
    "restore_focus",
  ];
  if (input.hasChildDeliveries) tail.push("queue_child_deliveries");
  if (input.hasReflectedUrl) tail.push("reflect_url");
  tail.push("dispatch_events", "run_registered_effects", "settle_feedback");

  if (input.render === "html") return ["preflight_morph", "morph", ...tail];
  if (input.render === "no_render") return ["validate_no_render", ...tail];
  return ["retain_dom", "settle_feedback"];
}

File: src/protocol/ordering.rs
//! Pure response-application planner with commit-after-morph semantics.

use super::{RenderPayload, ResponseOutcome, UpdateResponse, UpdateResponseV2, UrlIntent};
use crate::error::RecoveryInstruction;

/// Observed browser morph phase when planning an accepted HTML response.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum MorphDisposition {
    /// Morph has not run; used for redirects, no-render, and rejected outcomes.
    NotAttempted,
    /// Preflight and morph succeeded, so browser state may commit afterward.
    Succeeded,
    /// Server accepted but browser morph failed; original request must not replay.
    FailedAfterAcceptance,
}

/// One semantic application phase in the normative browser ordering model.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ApplicationStep {
    /// Perform real document navigation and no other response work.
    Navigate,
    /// Validate morph feasibility before touching current DOM.
    PreflightMorph,
    /// Reconcile the current island with server-rendered HTML.
    Morph,
    /// Validate the explicit no-render outcome in place of morphing.
    ValidateNoRender,
    /// Atomically install the accepted snapshot and browser revision.
    CommitSnapshotAndRevision,
    /// Reconcile model proposals and validation metadata.
    ReconcileModelsAndValidation,
    /// Restore focus according to the later browser policy.
    RestoreFocus,
    /// Queue signed parameter delivery through each surviving child scheduler.
    QueueChildDeliveries,
    /// Reflect accepted same-route state with `history.replaceState`.
    ReflectUrl,
    /// Dispatch declared events.
    DispatchEvents,
    /// Invoke registered effects without evaluating arbitrary script.
    RunRegisteredEffects,
    /// Settle loading, dirty, success, and error feedback.
    SettleFeedback,
    /// Keep existing DOM and browser authority unchanged.
    RetainDom,
    /// Obtain fresh authorized island state without replaying the accepted request.
    RequestFreshRenderWithoutReplay,
    /// Obtain fresh authorized island state for a classified server rejection.
    RequestFreshIsland,
    /// Stop Live processing for this boundary while preserving ordinary HTML.
    StopLive,
}

/// Produces the normative semantic application order for a fully parsed response.
#[must_use]
pub fn application_plan(
    response: &UpdateResponse,
    morph: MorphDisposition,
) -> Vec<ApplicationStep> {
    if response.redirect().is_some() {
        return vec![ApplicationStep::Navigate];
    }
    match response.outcome() {
        ResponseOutcome::Rejected => {
            return vec![ApplicationStep::RetainDom, ApplicationStep::SettleFeedback];
        }
        ResponseOutcome::RefreshRequired => {
            return if response
                .error()
                .is_some_and(|error| error.recovery() == RecoveryInstruction::Navigate)
            {
                vec![ApplicationStep::Navigate]
            } else {
                vec![
                    ApplicationStep::RetainDom,
                    ApplicationStep::RequestFreshIsland,
                    ApplicationStep::SettleFeedback,
                ]
            };
        }
        ResponseOutcome::Fatal => {
            return if response
                .error()
                .is_some_and(|error| error.recovery() == RecoveryInstruction::Navigate)
            {
                vec![ApplicationStep::Navigate]
            } else {
                vec![
                    ApplicationStep::RetainDom,
                    ApplicationStep::StopLive,
                    ApplicationStep::SettleFeedback,
                ]
            };
        }
        ResponseOutcome::Accepted | ResponseOutcome::Duplicate => {}
    }
    accepted_plan(response.render(), morph, false, false)
}

/// Produces protocol-v2 application order including child delivery and URL intent.
#[must_use]
pub fn application_plan_v2(
    response: &UpdateResponseV2,
    morph: MorphDisposition,
) -> Vec<ApplicationStep> {
    if response.redirect().is_some()
        || matches!(response.url_intent(), Some(UrlIntent::Navigated { .. }))
    {
        return vec![ApplicationStep::Navigate];
    }
    match response.outcome() {
        ResponseOutcome::Rejected => {
            return vec![ApplicationStep::RetainDom, ApplicationStep::SettleFeedback];
        }
        ResponseOutcome::RefreshRequired => {
            return if response
                .error()
                .is_some_and(|error| error.recovery() == RecoveryInstruction::Navigate)
            {
                vec![ApplicationStep::Navigate]
            } else {
                vec![
                    ApplicationStep::RetainDom,
                    ApplicationStep::RequestFreshIsland,
                    ApplicationStep::SettleFeedback,
                ]
            };
        }
        ResponseOutcome::Fatal => {
            return if response
                .error()
                .is_some_and(|error| error.recovery() == RecoveryInstruction::Navigate)
            {
                vec![ApplicationStep::Navigate]
            } else {
                vec![
                    ApplicationStep::RetainDom,
                    ApplicationStep::StopLive,
                    ApplicationStep::SettleFeedback,
                ]
            };
        }
        ResponseOutcome::Accepted | ResponseOutcome::Duplicate => {}
    }

    accepted_plan(
        response.render(),
        morph,
        !response.child_deliveries().is_empty(),
        matches!(response.url_intent(), Some(UrlIntent::Reflected { .. })),
    )
}

fn accepted_plan(
    render: Option<&RenderPayload>,
    morph: MorphDisposition,
    has_child_deliveries: bool,
    has_reflected_url: bool,
) -> Vec<ApplicationStep> {
    if matches!(render, Some(RenderPayload::Html(_)))
        && morph == MorphDisposition::FailedAfterAcceptance
    {
        return vec![
            ApplicationStep::PreflightMorph,
            ApplicationStep::Morph,
            ApplicationStep::RequestFreshRenderWithoutReplay,
        ];
    }

    let mut tail = vec![
        ApplicationStep::CommitSnapshotAndRevision,
        ApplicationStep::ReconcileModelsAndValidation,
        ApplicationStep::RestoreFocus,
    ];
    if has_child_deliveries {
        tail.push(ApplicationStep::QueueChildDeliveries);
    }
    if has_reflected_url {
        tail.push(ApplicationStep::ReflectUrl);
    }
    tail.extend([
        ApplicationStep::DispatchEvents,
        ApplicationStep::RunRegisteredEffects,
        ApplicationStep::SettleFeedback,
    ]);

    match render {
        Some(RenderPayload::Html(_)) => {
            let mut plan = vec![ApplicationStep::PreflightMorph, ApplicationStep::Morph];
            plan.extend(tail);
            plan
        }
        Some(RenderPayload::NoRender) => {
            let mut plan = vec![ApplicationStep::ValidateNoRender];
            plan.extend(tail);
            plan
        }
        None => vec![ApplicationStep::RetainDom, ApplicationStep::SettleFeedback],
    }
}
